cpp/include/cudf/detail/copy_if_else.cuh

```diff
@@ -71,23 +64,17 @@ __launch_bounds__(block_size) __global__
   size_type warp_cur = warp_begin + warp_id;
   size_type index    = tid;
   while (warp_cur <= warp_end) {
-    bool in_range = (index >= begin && index < end);
-
-    bool valid = true;
-    if (has_validity) {
-      valid = in_range && (filter(index) ? thrust::get<1>(lhs[index]) : thrust::get<1>(rhs[index]));
-    }
-
-    // do the copy if-else
-    if (in_range) {
-      out.element<T>(index) = filter(index) ? static_cast<T>(thrust::get<0>(lhs[index]))
-                                            : static_cast<T>(thrust::get<0>(rhs[index]));
+    bool valid = false;
+    if (index >= begin && index < end) {
+      auto value = filter(index) ? lhs[index] : rhs[index];
+      valid      = !has_validity || value.has_value();
+      if (valid) { out.element<T>(index) = static_cast<T>(value.value()); }
```

Comment:
The `valid` temporary is an unnecessary ITM. Also `optional` is implicit convertible to `bool`, so you can use it like a `ptr` when checking if it has a value or not. 
```suggestion
    if (index >= begin && index < end) {
      auto const value = filter(index) ? lhs[index] : rhs[index];
      if (not has_validity or value) { out.element<T>(index) = static_cast<T>(value.value()); }
```

Reply:
The `valid` variable is used in the next if-statement where the `warp_mask` is built.

Reply:
Okay, after actually looking at this, I think we can do better. First note that `index` is always greater or equal to `begin`, so the `index >= begin` is redundant. Also, lets rename `has_validity` to `has_nulls`. 

```suggestion
   auto const opt_value = (index < end) ? (filter(index) ? lhs[index] : rhs[index]) : thrust::nullopt;
   if(not has_nulls or opt_value){ out.element<T>(index) = *opt_value; }

   if(has_nulls){
      int warp_mask = __ballot_sync(0xFFFF'FFFF, opt_value.has_value();
```

Reply:
Cool. Thanks for looking at this. I'd prefer IIFE over nested ternary if that is ok.